synapse/lib/kv.py

```diff
@@ -0,0 +1,325 @@
+
+        Args:
+            prop (str): The property name.
+            valu (obj): A msgpack compatible value.
+        '''
+        lkey = self.iden + prop.encode('utf8')
+        self.stor.setKvProp(lkey, s_msgpack.en(valu))
+
+    def get(self, prop):
+        '''
+        Lookup a property from the KvLook.
+
+        Args:
+            prop (str): The property name.
+
+        Returns:
+            (obj): The return value, or None.
+        '''
+        lkey = self.iden + prop.encode('utf8')
+
+        lval = self.stor.getKvProp(lkey)
+        if lval is None:
+            return None
+
+        return s_msgpack.un(lval)
+
+class KvList:
+    '''
+    A KvList uses the KvStor to implement a pythonic list-like object.
+    '''
+    def __init__(self, stor, iden):
+        self.stor = stor
+        self.iden = iden
+
+        self.vals = [s_msgpack.un(b) for b in stor.iterKvDups(iden)]
+
+    def remove(self, valu):
+        '''
+        Remove a value from the KvList.
+
+        Args:
+            valu (obj): A msgpack value to remove.
+
+        Returns:
+            (bool): True if the value was found and removed.
+        '''
+        try:
+            self.vals.remove(valu)
+        except ValueError as e:
+            return False
+
+        lval = s_msgpack.en(valu)
+        self.stor.delKvDup(self.iden, lval)
+        return True
+
+    def __iter__(self):
+        return iter(self.vals)
+
+    def __len__(self):
+        return len(self.vals)
+
+    def append(self, valu):
+        '''
+        Append a value to the KvList.
+
+        Args:
+            valu (obj): A msgpack value to add.
+        '''
+        lval = s_msgpack.en(valu)
+
+        self.vals.append(valu)
+        self.stor.addKvDup(self.iden, lval)
+
+    def extend(self, vals):
+        '''
+        Extend the KvList by appending values from vals.
+
+        Args:
+            valu (list): A list of msgpack values to add.
+        '''
+        lval = [s_msgpack.en(v) for v in vals]
+
+        self.vals.extend(vals)
+        self.stor.addKvDups(self.iden, lval)
+
+class KvStor(s_eventbus.EventBus):
+    '''
+    The KvStor uses an LMDB key-value database to implement
+    persistence and indexing for primitive data structures.
+    '''
+
+    def __init__(self, path):
+        s_eventbus.EventBus.__init__(self)
+        self.lenv = lmdb.open(path, writemap=True, max_dbs=16)
+
+        self.oob = self.lenv.open_db(b'oob')
+        self.dups = self.lenv.open_db(b'dups', dupsort=True)
+
+        self.onfini(self._onKvFini)
+
+    def _onKvFini(self):
+        self.lenv.sync()
+        self.lenv.close()
+
+    @functools.lru_cache(maxsize=1024)
+    def genKvAlias(self, name):
+        '''
+        Resolve or create a new object alias by name.
+        '''
+        lkey = b'alias:' + name.encode('utf8')
+
+        with self.lenv.begin(db=self.oob, write=True) as xact:
+
+            iden = xact.get(lkey)
+            if iden is None:
+                iden = os.urandom(16)
```

Comment:
Is there a reason why we use `s_common.guid` for tufo idens and `os.urandom` for idens here?

Reply:
This is for raw random bytes,   s_common.guid returns a hex ascii string.